### ISSUE TITLE
Proposal: File scoped namespaces

### ISSUE THREAD

**Issue by MgSam** [dotnet/roslyn#595]
### Background
.NET style guidelines recommend having at most one namespace per file. The current namespace syntax only allows block scope, resulting in every file typically having all of its contents (save using statements) wrapped in a namespace block. This leads to both vertical space being wasted (the curly braces used to delimit the block) and more importantly, horizontal space being wasted, with all of the important content indented by an extra tab (if style guidelines are being followed). The indentation from the namespace block conveys no useful information- it is just code bloat.

``` C#
using System.Collections.Generic;

namespace Foo 
{
    public class Bar
    {
          ... // Potentially hundreds of lines of code, 
              // all of which has an extra indentation because 
              // of the namespace but is not actually conveying any useful information
    }
}
```

### Proposal
Allow a single namespace declaration per file that has no braces and therefore no block scope. Every class written in the file is a member of this namespace. It would be an error to attempt to introduce a second namespace declaration of any kind (including one with block scope). The single namespace declaration would be valid in the outermost scope only. It would be valid to appear at any part of the file in this scope, although style guidelines would recommend putting it at the top.

``` C# 
namespace Foo;

using System.Collections.Generic;

public class Bar
{
    ...
}
```

#### Previous Discussion
[Code Plex](https://roslyn.codeplex.com/discussions/544391)

**Comment by alanfo** [comment]
I think this is an excellent idea which should be very easy to implement.

Just one point.

As using directives can be local to a namespace, there would be a difference in meaning between:

```C#
namespace Foo;

using System.Collections.Generic;

public class Bar
{
    ...
}
```
and 

```C#
using System.Collections.Generic;

namespace Foo;

public class Bar
{
    ...
}
```
In the first case, the using statement would be local to Foo and, in the second case, it would be global .
 
However, there would be no practical difference between the two if no other namespaces were allowed in the file as specified in your proposal. 

In the interests of simplicity, it might therefore be best to require that the namespace declaration should always be the first line in the file.

**Comment by MgSam** [comment]
@alanfo The idea is that the namespace is file-scoped, so the question of whether a `using` statement is local to a namespace or not doesn't really apply. In your examples the behavior would be exactly the same. 

However, I do think that the rules for such a file-scoped namespace should probably mirror the rules for `using` statements. Thus, maybe it makes sense to require that the namespace is always the first non-comment, non-preprocessor line in the file.

**Comment by AdamSpeight2008** [comment]
What about the case of a partial class spread across multiple files, having different **file-scoped namespaces**

**Comment by RichiCoder1** [comment]
Wouldn't it ultimately end up meaning the same thing? File scoped just means the entire file lives in that namespace rather than allowing multiple.

**Comment by alanfo** [comment]
@MgSam 
Yes, I agree that it makes sense to require that the namespace is always the first non-comment, non-preprocessor line in the file.

@AdamSpeight2008 
The compiler would only combine partial classes with the same name if they were in the same namespace, whether file-scoped or not. They would be different classes otherwise.

@RichiCoder1 
Yes, it would be the same thing as putting the entire file within a single namespace. However, if the proposal were adopted, it would save you having to indent the file contents by an additional level. As a file having a single namespace is probably the commonest case, this would be a worthwhile improvement in my view.

**Comment by qrli** [comment]
This is what I hoped from version 1. it is not only wasting space, it also makes extra level of brace, where you'll need some IDE plugins make it easy to tell what each brace is for.
It was a copy of C++'s namespace but we know how combersome it is from the managed C++ experience.

**Comment by RichiCoder1** [comment]
Haha, sorry @alanfo, that was a response to @AdamSpeight2008. I absolutely agree.

**Comment by zgxnet** [comment]
Absolutely support! This is a very nice feature. It is really a waste of indentation to put every class inside a namespace scope. And it is an easy feature.

**Comment by mkosieradzki** [comment]
The indentation point is actually quite convincing. This is going to be a long shot: all this makes sense because statistically many developers actually have single namespace per file, but many developers also have single class per file and some nested classes. Maybe it's a good idea to allow similar syntax for classes then?

```
public sealed class Foo.Bar;

using System.Collections.Generic;

public sealed class MyNestedClass
{
    ...
}

public string Method1(int arg1)
{
     .....
}
```

or 

```
public namespace class Foo;
public sealed class Bar;

using System.Collections.Generic;

public sealed class MyNestedClass
{
    ...
}

public string Method1(int arg1)
{
     .....
}
```

**Comment by zgxnet** [comment]
Very nice proposal. I like the idea of file scope class. It will allow very clean code.

**Comment by tenor** [comment]
I don't see any real benefit to this proposal. It saves the developer a few keystrokes per file and removes an extra level of indentation (which any decent IDE can do automatically) but doesn't improve readability.

**Comment by paulomorgado** [comment]
Having something scoped to the compilation unit (file) is nothing new to **C#**. `using` directives can be scoped at the compilation unit level or namespace body level.

Because there are differences between compilation unit scoped and namespace scoped `using` directives, the proposal should account for that.

The [using keyword](https://msdn.microsoft.com/library/zhdeatwt.aspx "using (C# Reference)") is used for the [using directive](https://msdn.microsoft.com/library/sf0df423.aspx "using Directive (C# Reference)") and the [using statement](https://msdn.microsoft.com/library/yh598w02.aspx "using Statement (C# Reference)") but de difference is clear, not only because of where they are used, but how they are used. The using directive is a directive (there aren't many of them in **C#**) and has no body and the using statement has a body.

If fact, all statements introduced in **C#** (or not present is languanges commonly used at the time) that have associated statements/bodyes (`lock`, `using`, `get`, `set`) require them to be a block statement (with curly braces). I think they should have gone further and foreced that also for the others (`if`, `while`).

This proposal doesn't seem to fit well with **C#**.

If you really feel that those characters lost on the left are that important, I propose that the namespace declaration might be used without the curly braces, meaning that it will apply to whatever is next:

    namespace N1
        namespace N2
            class C1 {}

    namespace N1
        namespace N2
            class C2 {}

Or, beacause that space matters:

    namespace N1
    namespace N2
    class C1 {}

    namespace N1
    namespace N2
    class C2 {}

But, then, it becomes just about formatting (tooling):

    namespace N1
    {
    namespace N2
    {
    class C1 {}
    }
    }

    namespace N1
    {
    namespace N2
    {
    class C2 {}
    }
    }

Does that really worth all the hassle of the change?

How about this, instead?

    class N1.N2.C1 {}

    class N1.N2.C2 {}

**Comment by mkosieradzki** [comment]
I think @paulomorgado proposal is a nice compromise. It can be done, because namespaces do not have any modifiers what makes them different from classes, methods and so on.

**Comment by MgSam** [comment]
> Because there are differences between compilation unit scoped and namespace scoped using directives, the proposal should account for that.

I think this is an implementation detail. Is there a difference in behavior to the user? If not than it doesn't really matter which is used.

> If you really feel that those characters lost on the left are that important, I propose that the namespace declaration might be used without the curly braces, meaning that it will apply to whatever is next:

This would be both confusing and demonstrably less useful. Style guidelines recommend one class per file, but often you see files with more in them than just a single class. More than one namespace per class, on the other hand, is almost never seen.

> I don't see any real benefit to this proposal. It saves the developer a few keystrokes per file and removes an extra level of indentation (which any decent IDE can do automatically) but doesn't improve readability.

Visual Studio has certain defaults for files and auto-formatting behaviors it uses; C# has style guidelines which the vast majority of devs generally follow. All of these include a namespace enveloping the entirety of the file and indentation associated with that namespace. It's nice to talk about a hypothetical world in which this isn't a problem but the fact is the overwhelming majority code will continue to use the current style unless there's a compelling language-based reason not to.

**Comment by mkosieradzki** [comment]
For imperative-style it is a minor upgrade, but for functional-style it's a major upgrade. In imperative style you have a lot of short instructions one per line, control structures makes a lot of indents and outdents. In functional approach you have some definitions - sometimes one-per-line each consisting of usually long expression where IDE gives hand with utilities like word-wrap.

I think this couple of characters saved on the left will cause further word-wrapped lines to start earlier accumulating saved space on per-line basis. Consequently it means more actual lines visible per page and consequently more options to logically arrange code. It's far more than code will be more on the left.

**Comment by MgSam** [comment]
I added a link in the proposal to [UserVoice](http://visualstudio.uservoice.com/forums/121579-visual-studio/suggestions/7181921-file-scoped-namespaces), so you can vote for the feature if it interests you.

**Comment by paulomorgado** [comment]
@MgSam,

> > Because there are differences between compilation unit scoped and namespace scoped using directives, the proposal should account for that.
> 
> I think this is an implementation detail. Is there a difference in behavior to the user? If not than it doesn't really matter which is used.

§9.4.2 (Using namespace directives) does not mention the differences, but §7.6.5.2 (Extension method invocations) states that:

> Starting with the closest enclosing namespace declaration, continuing with each enclosing namespace declaration, and ending with the containing compilation unit, successive attempts are made to find a candidate set of extension methods:

This is valid for everyting namespace related. Using directives inside a namespace declaration superside any outer using directives.

So, this is not a style issue and, definetely, there's a difference to the user.

**Comment by tenor** [comment]
Hmm, nested namespaces brings up some interesting issues:

```C#
    namespace Namespace1;
    namespace Namespace2 {
        class C {}
    }
```
Does class C belong to `Namespace1.Namespace2` namespace or `Namespace2` namespace?
Should the compiler disallow other namespace declarations if a file scoped namespace is in use?

**Comment by MgSam** [comment]
@paulomorgado @tenor I think the same part of the spec I proposed answers both concerns:
> It would be an error to attempt to introduce a second namespace declaration of any kind (including one with block scope).

So binding extension methods is not an issue since all using statements would be within the same namespace scope. 

It would be unnecessarily confusing and complicating to allow namespace blocks to be declared when a file-scoped namespace is present.

**Comment by jnm2** [comment]
I like this idea. The forced amount of indentation has been a minor annoyance over the years, especially when copying. I'm not sure I'd be comfortable doing the same thing to the class declaration, but the beauty of this proposal is that it would work either way according to your preferences.

I have a possible alternative to the single line `namespace`. What about allowing namespaces in class names? This might look cleaner:

    using System;
    
    public class NameSpace.Product.ClassName
    {
        [...]
    }

**Comment by paulomorgado** [comment]
That was my proposal, @jnm2.

And it's not as strange as one might think because this:

    namespace N1
    {
        namespace N2
        {
            class C
            {
            }
        }
    }

is already the same as this:

    namespace N1.N2
    {
        class C
        {
        }
    }

**Comment by jnm2** [comment]
Oh, my bad. I'm 100% with you there.

**Comment by diryboy** [comment]
Nice, I like this. Just one max would be enough I think.

Some say, saving spaces improve compilation speed. I don't know, but it does improve readability, good for portrait screen coders, too.

**Comment by DavidArno** [comment]
Just to add my voice here, I'm in agreement with @MgSam: this is a simple proposal that brings some really nice readability benefits to code (especially when, as @mkosieradzki says, using a more functional style of development with C#).

Someone please shoot @paulomorgado though, for trying to unnecessarily complicate this proposal! :grinning:

**Comment by paulomorgado** [comment]
Ouch!

**Comment by jnm2** [comment]
It would be fantastic if I could specify both namespace and partial class:

```c#
public partial class MyNamespace.MyClass
{
    [...]
}
```

```c#
private class MyNamespace.MyClass.UtilityClass
{
    [...]
}
```

**Comment by anirudi** [comment]
Perfectly agree. It does not make sense to intend on namespace as anyways we can have only 1 namespace in a file.

**Comment by AdamSpeight2008** [comment]
We can as it is allowed by the C# grammar. Just because a majority of examples use only one, doesn't indicate that only one is allowed.

**Comment by int19h** [comment]
It's not just the majority of examples. It's a well-established code style guidelines, used by .NET Framework itself among other things.

If the de-facto official style guide results in unnecessarily verbosity (in this case, wasted indentation) for 99% of lines in every source file, something is clearly wrong.

One possibility is that IDE should simply never indent the body of the outermost `namespace` declaration. This is common in C++, but also fairly ugly.

Otherwise, I think the most reasonable approach is to come up with some simple syntactic sugar - in other words, whatever the syntax is, it should have a straightforward desugaring into the existing `namespace { ... }` declaration, and then all existing rules (for nested usings etc) should apply exactly as they do today.

Since there are already a few suggestions above, I'm going to add one more:
```cs
namespace Foo:
class Bar { ... }
```

So basically the same as the proposal, but using `:` to signify that everything that follows is the namespace body. This is consistent with `case ...:` in C#, and also with access specifiers in C++, so hopefully this syntax would reduce confusion.

There are only two simple rules for desugaring this:
1. `namespace X: ...` becomes `namespace X { ... }`, where `...` is everything after `:` up to the end of input.
2. `namespace X:` form is only permitted in global namespace - this prevents nesting. If you need to nest namespaces, use `{}` for inner ones.

This takes care of the 90% case (single namespace in a file) with straightforward, easy to understand syntax, and avoids adding new rules or creating corner cases.

**Comment by dsaf** [comment]
Frankly speaking, this feature would have more positive impact on an _average_ codebase than e.g. pattern matching. It's one file per type most of the time, and we remove 2-3 lines of code _and_ a whole indentation level.

**Comment by anirudi** [comment]
BMSEphx

Get Outlook for Android<https://aka.ms/ghei36>

**Comment by whoisj** [comment]
Frankly I'd take the ability to do class names long hand even.

```C#
public class Namespace.ClassName
{
    // define class here...
}
```

**Comment by alrz** [comment]
@MgSam I really like this feature, perhaps you want to port this to csharplang repo?

**Comment by MgSam** [comment]
@alrz Did they state they're not going to port the issues over and its up to the community?

**Comment by alrz** [comment]
> its up to the community?

@MgSam From what has happened so far, unless it's already championed, yes.

**Comment by jnm2** [comment]
@MgSam yes please port.

**Comment by AustinBryan** [comment]
Shouldn't style conventions prefer

```csharp

using System;

namespace MyNamspace;

public class MyClass { ... }
```

Over

```csharp

namespace MyNamspace;

using System;

public class MyClass { ... }
```
As it looks closer to what we're used to: 


```csharp
using System;

namespace MyNamespace
{
    public class MyClass { ... }
}
```

And we're already in the habit of checking the line right above the class name to find the namespace.

**Comment by paulomorgado** [comment]
What style conventions?
